alt-tab-macos/ui/PreferencesPanel.swift

```diff
@@ -1,110 +1,211 @@
+    }
+
+    private func makeLabelWithProvidedControl(_ labelText: String, _ rawName: String, _ control: NSControl, _ suffixText: String? = nil) -> NSStackView {
+        let label = NSTextField(wrappingLabelWithString: labelText + ": ")
+        label.alignment = .right
+        label.widthAnchor.constraint(equalToConstant: calcLabelWidth()).isActive = true
+
+        control.identifier = NSUserInterfaceItemIdentifier(rawName)
+        control.target = self
+        control.action = #selector(controlDidEndEditing)
+        let containerView = NSStackView(views: [label, control])
+
+        if suffixText != nil {
+            let suffix = NSTextField(labelWithString: suffixText!)
+            suffix.textColor = .gray
+            containerView.addView(suffix, in: .leading)
         }
+
+        return containerView
     }
 
-    func textDidEndEditing(_ notification: Notification) {
-        if let textView = notification.object as? NSTextView {
-            let key = inputsMap[textView]!
+    /*
+    usage notes:
+    - NSSlider: supports on purpose currently only decimal values
+    */
+    @objc func controlDidEndEditing(senderControl: NSControl) {
+        self.makeFirstResponder(invisibleTextField) // deselects any possibly selected NSTextField (so slider & popUp changes deselect them)
+
+        let key: String? = senderControl.identifier?.rawValue
+        var newValue: String?
+
+        if senderControl is NSPopUpButton {
+            newValue = (senderControl as! NSPopUpButton).titleOfSelectedItem!
+        } else if senderControl is NSSlider {
+            newValue = String(format: "%.0f", Double(senderControl.stringValue)!) // we are only interested in decimals of the provided double
+        } else {
+            newValue = senderControl.stringValue
+        }
+
+//        debugPrint("PreferencesPanel: save: change", key!, newValue!)
+
+        if key != nil && newValue != nil {
+            if newValue == Preferences.rawValues[key!] {
+                debugPrint("PreferencesPanel: save: abort: value was not changed")
+                return
+            }
+
+            let previousValue = Preferences.rawValues[key!]!
+
             do {
-                try Preferences.updateAndValidateFromString(key, textView.string)
+                try Preferences.updateAndValidateFromString(key!, newValue!)
                 try Preferences.saveRawToDisk()
             } catch {
-                debugPrint(key, error)
-                textView.string = Preferences.rawValues[key]!
+                debugPrint("PreferencesPanel: save: error", key!, error, "previousValue", previousValue, " | newValue", newValue!)
+
+                // restores the previous value in Preferences and senderControl
+                try! Preferences.updateAndValidateFromString(key!, previousValue)
+
+                if senderControl is NSPopUpButton {
+                    (senderControl as! NSPopUpButton).selectItem(withTitle: previousValue)
+                } else {
```

Comment:
This handles a runtime use case: to reset the corresponding `Preferences` memory value and `PreferencesPanel` view value to its `previousValue` when there was an issue with storing `newValue` into the Preferences (saveOnDisk). 

Since we currently do not inform the user about such things (notification/alert/modal missing) I think it made sense to recover from that not only in the Preferences memory values but also visually. The user will see there something strange going on and not expect his `newValue` to be actually set. In my opinion we need something like this. You could simulate this use-case by making the .json file to be read-only (some disk or permissions issue). 

Once we implement #73 we will have a way to show alerts of some kind which could then be used here as well to show a message for the user in addition to the revert of visual & memory values.

Reply:
You were previously changing the views value in a catch block. I was thinking about your motivation behind that code line 👍 and interpreted that use case.

```
            do {
                try Preferences.updateAndValidateFromString(key, textView.string)
                try Preferences.saveRawToDisk()
            } catch {
                debugPrint(key, error)
                textView.string = Preferences.rawValues[key]!
            }
```

Reply:
My point is that the code you quoted from before made sense when the user is facing an input field. They can type "aaa" in the max window input field for instance, and when they do that we want to reset it to the previous value as cheap way to handle input field validation.

For dropdowns and sliders, the user can't select a wrong value anymore as we can make sure all finite values are valid at compile time. Thus this code is no longer needed for these finite inputs. Imagine the slider being moved and then that code moving it back to the previous value on release. This is a logical bug, not input validation

Reply:
Only now I get what this was for. Good we discussed that one.

> Imagine the slider being moved and then that code moving it back to the previous value on release. This is a logical bug, not input validation

My intention was not input validation but handling an issue while saving Preferences (e.g. disk issue or whatever else could go wrong).

Ive misinterpreted your intention behind that code line and didn't expect this to be a smart way for input validation 👍 

Ok I've gonna throw an NSAlert to inform the user about an issue. Is it ok to leave the keep the input values alone then (not reverting at all)?

Reply:
Ok, ive figured out the NSAlert. It gives the user 3 choices: 

1. Ok (ignoring the failure, not touching the controls value)
2. Try again (repeats `controlDidEndEditing`)
3. Revert value (to `previousValue`)